code/controllers/subsystem/tts.dm

```diff
@@ -0,0 +1,227 @@
+		SEND_SOUND(target, sound)
+		return
+
+	var/turf/turf_source = get_turf(target)
+	if(!turf_source)
+		return
+
+	var/channel = SSsounds.random_available_channel()
+	var/listeners = get_hearers_in_view(range, turf_source)
+
+	for(var/mob/listening_mob in listeners | SSmobs.dead_players_by_zlevel[turf_source.z])//observers always hear through walls
+		var/datum/language_holder/holder = listening_mob.get_language_holder()
+		if(!listening_mob.client?.prefs.read_preference(/datum/preference/toggle/sound_tts))
+			continue
+
+		if(get_dist(listening_mob, turf_source) <= range && holder.has_language(language, spoken = FALSE))
+			listening_mob.playsound_local(
+				turf_source,
+				sound,
+				vol = listening_mob == target? 60 : 85,
+				falloff_exponent = SOUND_FALLOFF_EXPONENT,
+				channel = channel,
+				pressure_affected = TRUE,
+				sound_to_use = sound,
+				max_distance = SOUND_RANGE,
+				falloff_distance = SOUND_DEFAULT_FALLOFF_DISTANCE,
+				distance_multiplier = 1,
+				use_reverb = TRUE
+			)
+
+/datum/controller/subsystem/tts/proc/handle_request(list/entry)
+	var/timeout_time = entry[START_TIME_INDEX] + message_timeout
+	if(timeout_time < world.time)
+		cached_voices -= entry[IDENTIFIER_INDEX]
+		return
+	var/datum/http_request/request = entry[REQUEST_INDEX]
+	request.begin_async()
+	in_process_tts_messages += list(entry)
+
+/datum/controller/subsystem/tts/fire(resumed)
+	if(!tts_enabled)
+		flags |= SS_NO_FIRE
+		return
+
+	if(!resumed)
+		while(length(in_process_tts_messages) < max_concurrent_requests && queued_tts_messages.len > 0)
+			var/list/entry = popleft(queued_tts_messages)
+			handle_request(entry)
+		current_processing_tts_messages = in_process_tts_messages.Copy()
+
+	// For speed
+	var/list/processing_messages = current_processing_tts_messages
+	while(processing_messages.len)
+		var/current_message = processing_messages[processing_messages.len]
+		processing_messages.len--
+		var/datum/http_request/request = current_message[REQUEST_INDEX]
+		if(!request.is_complete())
+			continue
+
+		var/datum/http_response/response = request.into_response()
+		in_process_tts_messages -= list(current_message)
+		average_tts_messages_time = MC_AVERAGE(average_tts_messages_time, world.time - current_message[START_TIME_INDEX])
+		// If it took too long to process, don't bother playing it
+		var/timeout_time = current_message[START_TIME_INDEX] + message_timeout
+		if(response.errored || timeout_time < world.time)
+			cached_voices -= current_message[IDENTIFIER_INDEX]
+			continue
+
+		var/identifier = current_message[IDENTIFIER_INDEX]
+		var/sound/new_sound = new("tmp/tts/[identifier].ogg")
+		play_tts(current_message[TARGET_INDEX], new_sound, current_message[LANGUAGE_INDEX], current_message[LOCAL_INDEX], current_message[MESSAGE_RANGE_INDEX])
+		for(var/extra_target in current_message[EXTRA_TARGETS_INDEX])
+			play_tts(extra_target["target"], new_sound, current_message[LANGUAGE_INDEX], extra_target["local"], extra_target["range"])
```

Comment:
what's in this list exactly? why does it exist?

Reply:
extra targets, who have tried to request the same tts message, with the same voice and same filters.
![image](https://user-images.githubusercontent.com/37270891/236446999-607626af-9fd2-4cbc-bc43-663583edf283.png)
This is where cached_voice comes in useful since it means I don't need to send a double request to the http server to generate the same voice clip.

Reply:
when would this happen? two people hearing the same guy? or the same guy saying something twice.
also uh, why not just have a list of targets? why do it as a sublist like this seems sorta off

Reply:
targets are the people who are saying the TTS message. This is unlikely to be triggered in most cases but it'll get triggered if some guy says the same message with the same voice whilst as another if the other person's tts message is still in processing.
It's just a way of reducing the number of tts requests sent to the tts server since duplicate requests won't be sent

Reply:
it's odd to me that there's a set of vars on the voice that describe one target, and then any further targets get put into a special sublist.
It feels like there should just be a list of targets, and that's it.

Reply:
is this blocking @LemonInTheDark

Reply:
that makes sense to me